server/blob/blob-cassandra/src/main/java/org/apache/james/blob/cassandra/cache/CachedDumbBlobStore.java

```diff
@@ -0,0 +1,80 @@
+package org.apache.james.blob.cassandra.cache;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+
+import javax.inject.Inject;
+
+import org.apache.james.blob.api.BlobId;
+import org.apache.james.blob.api.BucketName;
+import org.apache.james.blob.api.DumbBlobStore;
+import org.apache.james.blob.api.ObjectNotFoundException;
+import org.apache.james.blob.api.ObjectStoreIOException;
+import org.reactivestreams.Publisher;
+
+import com.github.fge.lambdas.Throwing;
+import com.google.common.base.Preconditions;
+import com.google.common.io.ByteSource;
+
+import reactor.core.publisher.Mono;
+
+public class CachedDumbBlobStore implements DumbBlobStore {
+
+    private final DumbBlobStoreCache cache;
+    private final DumbBlobStore backend;
+
+    @Inject
+    public CachedDumbBlobStore(DumbBlobStoreCache cache, DumbBlobStore backend) {
+        this.cache = cache;
+        this.backend = backend;
+    }
+
+    @Override
+    public InputStream read(BucketName bucketName, BlobId blobId) throws ObjectStoreIOException, ObjectNotFoundException {
+        Preconditions.checkNotNull(bucketName, "bucketName should not be null");
+
+        return Mono.from(cache.read(blobId))
+            .map(bytes -> (InputStream) new ByteArrayInputStream(bytes))
+            .switchIfEmpty(Mono.fromCallable(() -> backend.read(bucketName, blobId)))
+            .blockOptional()
+            .orElseThrow(() -> new ObjectNotFoundException(String.format("Could not retrieve blob metadata for %s", blobId)));
+    }
+
+    @Override
+    public Publisher<byte[]> readBytes(BucketName bucketName, BlobId blobId) {
+        return Mono.from(cache.read(blobId))
+            .switchIfEmpty(Mono.from(backend.readBytes(bucketName, blobId)));
+    }
+
+    @Override
+    public Publisher<Void> save(BucketName bucketName, BlobId blobId, byte[] data) {
+        return Mono.from(cache.cache(blobId, data))
+            .then(Mono.from(backend.save(bucketName, blobId, data)));
+    }
+
+    @Override
+    public Publisher<Void> save(BucketName bucketName, BlobId blobId, InputStream inputStream) {
+        return Mono.fromCallable(() -> inputStream)
+            .map(stream -> cache.cache(blobId, stream))
```

Comment:
here the `stream` is consumed, does it throw if it is read again?

Reply:
Yes that let me think the small stream handling should be here and not in the cache.

Reply:
Save in the cache only for default bucket

Reply:
> The stream is pass to cache and backend, in these method, stream are copied to another stream and process it, so it not throw ioeException.

Not sure to understand but trying to copy inside cache and/or backend is a dead end: it will be consumed for the other implementation.

The logic about threshold should be handled here as suggested by @chibenwa and we should only use the cache method with a `byte[]`

Reply:
So, the picture is clear.Should we fix at https://github.com/linagora/james-project/pull/3273/